Mathlib/Data/Multiset/Basic.lean

```diff
@@ -2316,6 +1998,297 @@ theorem filter_eq' (s : Multiset α) (b : α) : s.filter (· = b) = replicate (c
+protected def sub (s t : Multiset α) : Multiset α :=
+  (Quotient.liftOn₂ s t fun l₁ l₂ => (l₁.diff l₂ : Multiset α)) fun _v₁ _v₂ _w₁ _w₂ p₁ p₂ =>
+    Quot.sound <| p₁.diff p₂
+
+instance : Sub (Multiset α) :=
+  ⟨Multiset.sub⟩
+
+@[simp]
+lemma coe_sub (s t : List α) : (s - t : Multiset α) = (s.diff t : List α) :=
+  rfl
+
+/-- This is a special case of `tsub_zero`, which should be used instead of this.
+  This is needed to prove `OrderedSub (Multiset α)`. -/
+@[simp, nolint simpNF] -- We want to use this lemma earlier than the lemma simp can prove it with
+protected lemma sub_zero (s : Multiset α) : s - 0 = s :=
+  Quot.inductionOn s fun _l => rfl
+
+@[simp]
+lemma sub_cons (a : α) (s t : Multiset α) : s - a ::ₘ t = s.erase a - t :=
+  Quotient.inductionOn₂ s t fun _l₁ _l₂ => congr_arg _ <| diff_cons _ _ _
+
+protected lemma zero_sub (t : Multiset α) : 0 - t = 0 :=
+  Multiset.induction_on t rfl fun a s ih => by simp [ih]
+
+@[simp]
+lemma countP_sub {s t : Multiset α} :
+    t ≤ s → ∀ (p : α → Prop) [DecidablePred p], countP p (s - t) = countP p s - countP p t :=
+  Quotient.inductionOn₂ s t fun _l₁ _l₂ hl _ _ ↦ List.countP_diff hl _
+
+@[simp]
+lemma count_sub (a : α) (s t : Multiset α) : count a (s - t) = count a s - count a t :=
+  Quotient.inductionOn₂ s t <| by simp [List.count_diff]
+
+/-- This is a special case of `tsub_le_iff_right`, which should be used instead of this.
+This is needed to prove `OrderedSub (Multiset α)`. -/
+protected lemma sub_le_iff_le_add : s - t ≤ u ↔ s ≤ u + t := by
+  revert s
+  exact @(Multiset.induction_on t (by simp [Multiset.sub_zero]) fun a t IH s => by
+      simp [IH, erase_le_iff_le_cons])
+
+protected lemma sub_le_self (s t : Multiset α) : s - t ≤ s := by
+  rw [Multiset.sub_le_iff_le_add]
+  exact le_add_right _ _
+
+instance : OrderedSub (Multiset α) :=
+  ⟨fun _n _m _k => Multiset.sub_le_iff_le_add⟩
+
+instance : ExistsAddOfLE (Multiset α) where
+  exists_add_of_le h := leInductionOn h fun s =>
+      let ⟨l, p⟩ := s.exists_perm_append
+      ⟨l, Quot.sound p⟩
+
+@[simp]
+lemma filter_sub (p : α → Prop) [DecidablePred p] (s t : Multiset α) :
+    filter p (s - t) = filter p s - filter p t := by
+  revert s; refine Multiset.induction_on t (by simp) fun a t IH s => ?_
+  rw [sub_cons, IH]
+  by_cases h : p a
+  · rw [filter_cons_of_pos _ h, sub_cons]
+    congr
+    by_cases m : a ∈ s
+    · rw [← cons_inj_right a, ← filter_cons_of_pos _ h, cons_erase (mem_filter_of_mem m h),
+        cons_erase m]
+    · rw [erase_of_not_mem m, erase_of_not_mem (mt mem_of_mem_filter m)]
+  · rw [filter_cons_of_neg _ h]
+    by_cases m : a ∈ s
+    · rw [(by rw [filter_cons_of_neg _ h] : filter p (erase s a) = filter p (a ::ₘ erase s a)),
+        cons_erase m]
+    · rw [erase_of_not_mem m]
+
+@[simp]
+lemma sub_filter_eq_filter_not (p) [DecidablePred p] (s : Multiset α) :
+    s - s.filter p = s.filter fun a ↦ ¬ p a := by
+  rw [← filter_add_not p s]
+  ext a; by_cases h : p a <;> simp [h]
```

Comment:
If this proof is already multiline, might as well go all the way. (Also: why is this extra `rw` needed?)
```suggestion
  rw [← filter_add_not p s]
  ext a
  by_cases h : p a <;> simp [h]
```

Reply:
Not sure why that `rw` ended up here. I've removed it (and the proof is now back to one line)